### DISCUSSION TITLE
NextAuth with NextJS + external backend authentication

### DISCUSSION THREAD

**Discussion by xerai** [nextauthjs/next-auth#8005] [Help]
I'm trying to implement NextAuth with Google provider login so that after login I could also authenticate the logged in user on my external Go backend where from I would try to fetch user's data etc.

I have tried to configure it so that it puts encrypted JWT (=JWE) into cookies on every request and uses the same key on Nextjs and Go backend when encrypting and decrypting.

However I haven't managed to decrypt and validate/authenticate the user successfully on backend due to the decryption fails every time and I would like to confirm if I have configured the NextAuth correct.

So here is my pages/api/auth/[...nextauth.js]:
```
export const authOptions = {
	encryption: true,
	signingKey: process.env.NEXTAUTH_SECRET,
  providers: [
	GoogleProvider({
		clientId: process.env.GOOGLE_CLIENT_ID,
		clientSecret: process.env.GOOGLE_CLIENT_SECRET,
	  }),
  ],
  jwt: {
	signingKey: process.env.NEXTAUTH_SECRET,
	encryption: true,
  },
  session: {
	strategy: "jwt",
	maxAge: 30 * 24 * 60 * 60, // 30 days
	jwt: true,
	},

  callbacks: {
	async jwt({ token, account, profile }) {
		// Persist the OAuth access_token and or the user id to the token right after signin
		if (account) {
		  token.accessToken = account.access_token
		  token.id = profile.id
		}
		return token
	},
	async session({ session, token}) {
		// Send properties to the client, like an access_token and user id from a provider.
		session.accessToken = token.accessToken
		session.token = token
		return session
	  }
  },
}
```

And after login via NextAuth we return to index page and try to fetch user notes as:
```
const fetchNotes = async () => {
    if (!session) return;
    const response = await fetch("http://localhost:8080/notes", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
      credentials: "include",
    });
    const data = await response.json();
    setNotes(data.notes);
  };
```

So the backend receives for example this kind of value as `cookie.Value`:
`eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0XNIn0..XELY7eqL0LDsFS0-.L8uGU8XDBJMX4t6ZnSh3fS6p_vRIF3k8YHnB2lPkxbgVHXkFKeSNmWNRKMf41KjejAXQaoTC4KGsBACLuP4LtA0KmNxjTkf8CD9fyt4uLdgHg3GJghXqPceXX0qqtwupkCK5pvMc2lcoicIS1P3vdF_j7uCJavOF5FXj39fenZEzVw53Q1ouqGrwQghjJFYfFJVDa5Ik9rNljlw6ZEt9NvTqwzEdoFVJPLfmt885-yrSzkPdSGNcR2YqqmrKNjkuoAYt4OI0AkyD5EGPZkUIJgKlDcttgbPfTPaygEqpTAFzeIJilJI3cT2It5NTLx0O_bpeXJJmaamRQN4xPcRn1zm9V7x7D35Uj12qq8me-6tjmqSCz6Y959OqeNuuaVRJVjG8akEw_mnyugu572YUL9pSQCegO7Ih722Y8mPbJTFzs_3VmTQx8Z704HJMSDdU_6fShd_mCeVL6TJ6E8zR2zKUNm_WQW9u8yoFRcufaAIQrYzKRb6WqoUNrYaJ5VsDZAGOe152YaBvibrxRHFEh3BwlKwwRibgjbXzeJvs2kFxFbVggJmbM0O9_9UAAvkqYJ4MTYgZx1iMClK3Y6CnZEErDgBM15rdT3xzouYO8IhTUdw-jcCgq-lGc9yPvcKfZpA5Lmm0FzpfNCd7Ow.KLwTXzWunR4roYZp8J9BKw `

Then on my Go backend I'm trying to decrypt the JWE but it fails on different librarys. Here is one example:
```
func Auth() gin.HandlerFunc {
    return func(context *gin.Context) {
		//fmt.Println("context request", context.Request)
        base64Key := os.Getenv("SECRET_KEY")
        key, err := base64.StdEncoding.DecodeString(base64Key)
        if err != nil {
            panic(err)
        }

        cookie, err := context.Request.Cookie("next-auth.session-token")
        fmt.Println("cookie.val", cookie.Value, "err", err)
        if err != nil {
            context.JSON(401, gin.H{"error": "request does not contain a JWT cookie"})
            context.Abort()
            return
        }
        jwe, err := jose.ParseEncrypted(cookie.Value)
        if err != nil {
            panic(err)
        }

        decrypted, err := jwe.Decrypt(key)
        if err != nil {
            panic(err)
        }

        fmt.Printf("Decrypted JWE: %s", decrypted)

    }

}
```
Error: `square/go-jose: error in cryptographic primitive`

That error usually hints that encryption/decryption is not done with same key, or there is invalid value.
I have checked that the Go backend uses same key as the Next.js where I have defined the NEXTAUTH_SECRET on .env.local and key made as the docs suggested.

So does someone have some idea where this goes wrong? Any help would be appreciated.

**Comment by getellez** [comment]
Did you make sure the secret is 32 bytes long?
